### DISCUSSION TITLE
wasmCloud Q1 2025 Roadmap

### DISCUSSION THREAD

**Discussion by brooksmtownsend** [wasmCloud/wasmCloud#3813] [Community RFC]
Hello wasmCloud Community!

Every quarter we publish roadmaps for wasmCloud after community planning sessions during our weekly Wednesday community meeting. [wasmcloud.com/docs/roadmap](https://wasmcloud.com/docs/roadmap/) From some community feedback, we're opening this discussion a little earlier to allow for more time for feature requests, bug reports, big ideas, and more.

This discussion is for suggesting features, prioritizing fixes, and general brainstorming for the priorities of wasmCloud maintainers + contributors going into Q1 of next year. Our previous roadmap focused on community contributions, improving developer workflows, and finalizing some support for WASI interfaces. Based on the discussion here, we'll determine our next roadmap's focus.

Please feel free to comment here whether you're a long-time wasmCloud user, maintainer, first-time contributor, or just curious onlooker that noticed something missing!

**Comment by ossfellow** [comment]
As a community member and contributor, I see wasmCloud platform and community growing in multiple dimensions, one of which is the gradual maturation and deepening of the core platform, its features and capabilities. In addition, wasmCloud is a major contributor to the WASI component model, which benefits the greater Wasm/WASI ecosystem, and on top of that now a CNCF incubating project 🎉

So, with all that in mind, where I find my contribution more impactful is in implementing WASI interface proposals, in the form of wasmCloud capability providers, and good examples, which make the platform more useful and accessible to the new and existing end users and fellow community members.

With artificial intelligence pushed to the forefront, combined with availability of wasmCloud on a wide range platforms, from cloud all the way to the [user edge](https://lfedge.org/breaking-down-the-edge-continuum/), and [wRPC transport](https://github.com/bytecodealliance/wrpc) to connect it all, in this quarter, I'd like to implement the [wasi-nn](https://github.com/WebAssembly/wasi-nn) proposal, in 2 phases (i.e. 2 tickets):

1. wasmCloud runtime bindings for wasi-nn, so maintainers could evaluate the accuracy and robustness of the implementation.
2. Building on number 1, the wasi-nn capability provider, and its associated integration test suite.

I also like to implement the [watcher interface](https://github.com/WebAssembly/wasi-keyvalue/blob/main/wit/watch.wit), for the [NATS key-value store provider](https://github.com/wasmCloud/wasmCloud/tree/main/crates/provider-keyvalue-nats), and add a test suite for the provider.

Finally, I'd like to improve/fix handling of 2 edge cases (**copy_object** and **move_object**) in [Azure](https://github.com/wasmCloud/wasmCloud/tree/main/crates/provider-blobstore-azure), [Filesystem](https://github.com/wasmCloud/wasmCloud/tree/main/crates/provider-blobstore-fs), and [S3](https://github.com/wasmCloud/wasmCloud/tree/main/crates/provider-blobstore-s3), Blobstore providers. Their tickets don't necessarily need to be part of the Q1 2025 roadmap, but I'd like to have their tickets to ensure they will be covered.

What I like to see?

- The [documentation for handling of custom WIT resources](https://github.com/wasmCloud/wasmcloud.com/issues/639), which was added in Q4 2024 roadmap, as an enabler of building real-life solutions on top of wasmCloud platform.
- Fixing and improving the testing framework, especially given the higher profile of the platform as an incubating project.

**Comment by ffuerste** [comment]
As a member of the Community, I would like to propose the following topics and priorities:

## Autocompletion for `wash app` subcommands

* continuation of the [already started implementation](https://github.com/wasmCloud/wasmCloud/issues/3258)
* support autocompletion for `wash deploy` for the respective application manifest file (similar to `kubectl apply -f <filename>`)
* support of an application manifest as source for `wash app undeploy` (similar to `kubectl delete -f <filename>`), including autocompletion for the filename

I think this would significantly increase user-friendliness and, above all, make manual testing much easier, where applications have to be deployed and undeployed again and again. Automated testing could also benefit from supporting application manifests as source, removing a potential user error where the application name in the manifest and for the `wash app undeploy` command are not the same.

## Provenance and Attestation for wasmCloud Core Components

* Processing of the [started RFC](https://github.com/wasmCloud/wasmCloud/issues/3787)
* implementation of the discussed and aligned steps/measures for the wasmCloud core components, e.g.
  * Github action pipeline
  * user documentation

While all of the benefits and motivations already mentioned in the [corresponding RFC]() are important, the most important point for me is that it enables the foundation for enterprise-level security, which is imperative for regulated and/or security-focused organizations. Especially for companies that participate in audits such as the

* [C5](https://www.bsi.bund.de/EN/Themen/Unternehmen-und-Organisationen/Informationen-und-Empfehlungen/Empfehlungen-nach-Angriffszielen/Cloud-Computing/Kriterienkatalog-C5/C5_Einfuehrung/C5_Einfuehrung_node.html)
* [EUCS](https://ec.europa.eu/newsroom/cipr/items/713799/en)

## Updating User Documentation

* Updating the [`link` trait](https://wasmcloud.com/docs/ecosystem/wadm/model#links) documentation
  * The deprecated fields `source_config` and `target_config` are currently still documented.
  * we need to replace them with the new fields `source` and `destination` and (maybe) a short description of why the deprecated fields still work
* Discussion/introduction of a formal process to ensure that
  * user documentation is updated together with the respective code changes (if such a process does not yet exist)
  * Spreading knowledge about (breaking) changes and how to deal with it (or its effects)

WasmCloud is a constantly evolving project that will continue to introduce breaking changes in the future and we need to ensure that the user documentation is up to date.

**Reply by brooksmtownsend** [reply]
> Autocompletion for wash app subcommands

Completely agree here, this feature would make using the wash CLI much less painful! Swapping back to the mouse for copy/pasting all the time is tedious.

One note here, you can currently use the application manifest as an argument to undeploy!

```bash
➜ wash app undeploy app.yaml

Undeployed application: go-template
```

**Reply by brooksmtownsend** [reply]
> user documentation is updated together with the respective code changes (if such a process does not yet exist)

I am really interested in trying to approach this first and foremost with autogenerated documentation wherever possible. Event payloads, API changes, even command line args should all be something we can autogenerate and update when the code is updated. I think it's easier said than done of course, but we should try and remove the human element where possible!

@ericgregory do you have any good ideas / processes that you've seen before when it comes to making sure new code changes update the documentation as well? Our documentation is extensive enough now that it's hard to know when a change is going to affect it

**Comment by markkovari** [comment]
I personally (or a community member 😄 ) would be happy to see the followings being implemented or proceeded in any way:

1. The support of multitenancy in wasmCloud (there has been [PR](https://github.com/wasmCloud/wasmCloud/pull/3313)/[RFC](https://github.com/wasmCloud/wasmCloud/issues/3120)  about this)
2. An implementation of a scheduler provider (or even a built-in one) - which already has [an issue](https://github.com/wasmCloud/wasmCloud/issues/3821)

beside these I do support everything the others mention above. 🎉

**Reply by brooksmtownsend** [reply]
That sounds great! @markkovari do you have any additional feedback on multitenancy or what you would use the feature for? I'd love to mark the multitenancy RFC as something we'd implement in Q1, but I fear that it may not be completed. I think we might need more information behind that RFC before we go to implementation, and I'll call this out in the roadmapping session on Wed.

**Reply by markkovari** [reply]
Yeah, that part was a bit more over-enthusiastic from my side too, what I'll definitely do is to writeup the use case that I see/would expect.

But thanks for the heads up, I am usually jumping to conclusions early.

**Comment by Aditya1404Sal** [comment]
similiar to what @markkovari mentioned
I am extremely interested in implementing a built-in cron job provider with respect to the updated wasmcloud host, I understand that there is a pre-existing cronjob provider compliant with an older IDL. As a fairly recent community member, I want to work on a periodic trigger mechanism that's up-to-date with WIT.

Other than this I'm still exploring wasmcloud and I'd love to advocate this project by creating videos that would function as a step-by-step guide for helping users understand the development process behind wasm components.

**Reply by brooksmtownsend** [reply]
I love the idea of additional videos for tutorials 😄

**Comment by brooksmtownsend** [comment]
Feedback from here officially incorporated in the roadmap in the 15 Jan 2025 community meeting https://wasmcloud.com/community/2025/01/15/community-meeting
